spylib/token.py

```diff
@@ -0,0 +1,428 @@
+                rate = jsondata['extensions']['cost']['throttleStatus']["restoreRate"]
+                sleep_time = ceil((query_cost - available) / rate)
+                await sleep(sleep_time)
+                raise ShopifyThrottledError
+            elif OPERATION_NAME_REQUIRED_ERROR_MESSAGE in errorlist:
+                raise ShopifyCallInvalidError(
+                    f'Store {self.store_name}: Operation name was required for this query.'
+                    'This likely means you have multiple queries within one call '
+                    'and you must specify which to run.'
+                )
+            elif WRONG_OPERATION_NAME_ERROR_MESSAGE.format(operation_name) in errorlist:
+                raise ShopifyCallInvalidError(
+                    f'Store {self.store_name}: Operation name {operation_name}'
+                    'does not exist in the query.'
+                )
+            else:
+                raise ValueError(f'GraphQL query is incorrect:\n{errorlist}')
+
+        return jsondata['data']
+
+    def __new__(cls, *args, **kwargs):
+        """
+        Function to keep user from instantiating an instance of this class.
+        """
+        if cls is Token:
+            raise TypeError(f'Can\'t instantiate abstract class {cls.__name__} directly')
+        return object.__new__(cls)
+
+    def save_json(self):
+        json.dumps(
+            {
+                'store_name': self.store_name,
+                'scope': self.scope,
+                'access_token': self.access_token,
+                'access_token_invalid': self.access_token_invalid,
+                'client': self.client,
+                'oauth_url': self.oauth_url,
+                'api_url': self.api_url,
+                'api_version': self.api_version,
+                'rest_bucket_max': self.rest_bucket_max,
+                'rest_leak_rate': self.rest_leak_rate,
+                'graphql_bucket_max': self.graphql_bucket_max,
+                'graphql_leak_rate': self.graphql_leak_rate,
+            }
+        )
+
+
+class OfflineTokenABC(Token, ABC):
+    """
+    Offline tokens are used for long term access, and do not have a set expiry.
+    """
+
+    def __init__(self, **kwargs) -> None:
+        super().__init__(**kwargs)
+
+    async def reset_token(self, client_id: str, client_secret: str, code: str):
+
+        token: OfflineTokenResponse = OfflineTokenResponse(
+            **await self._get_token(client_id, client_secret, code)
+        )
+        self.access_token = token.access_token
+        self.scope = token.scope.split(',')
+
+
+@dataclass
+class OnlineTokenABC(Token):
+    """
+    Online tokens are used to implement applications that are authenticated with
+    a specific user's credentials. These extend on the original token, adding
+    in a user, its scope and an expiry time.
+    """
+
+    associated_user_scope: Optional[List[str]]
```

Comment:
You already have `scope` in the `Token` class.
```suggestion
```

Reply:
I was reading the docs and they say that the user can have a subset of the app scopes for online tokens, which is returned in the token response. This is what the OAuth endpoint returns:

https://github.com/SatelCreative/satel-spylib/blob/e4375f8a6a072fec2f36a4631b0cf7389806ed9a/spylib/token.py#L52-L57


Should I leave it?

Reply:
Oh I didn't realize that Shopify sends the app scope AND the user scope for online tokens ... I thought it'd send only the online token one.
I don't see the point of `scope` here since what any online token will be able to do is what is allowed by the `associated_user_scope` ...
It also means that any code we write that uses the scope will have to be different for online and offline tokens. Sounds annoying. It seems to me it would be better to save `associated_user_scope` in our `scope`. Maybe we should rename the `scope` field to something like `token_scope` to clarify that it's not the app scope or the user scope, but the token scope so it depends on the type of token ...
We can use pydantic's aliases and have a different one for Online and Offline tokens. It's not pretty but that's what we have.

Reply:
I think they do this because you are supposed to verify that the user scope that is returned is a subset of the app scope. You can re-request the auth for the user if the new scope of the application is less than the original scope for the user.

Should we add in the check as middleware so on every request using online tokens it checks to be sure that the token scopes are valid?

Reply:
The validity of the scopes should be checked during the online token oauth process. That's enough. Once it was allowed and we stored it, it's good to go.

Reply:
So it is safe to assume that the app scope will never change? I'm just thinking if the apps scope becomes more restrictive and removes an existing permission from an OnlineToken, there will be no way to know/renew this token. We will just have to dump and renew all tokens forcefully or users will run into weird errors they will have no control over.

Reply:
We currently have no process for the renewal anyway. We need to do that. That's another item that belongs in another issue/PR. Again we can add it later if needed but let's not over-engineer this.